crates/ruff_python_parser/src/semantic_errors.rs

```diff
@@ -17,6 +19,11 @@ use ruff_text_size::{Ranged, TextRange, TextSize};
 use rustc_hash::FxHashSet;
 
 #[derive(Debug)]
+pub struct Checkpoint {
+    in_async_context: bool,
+}
```

Comment:
We could potentially include a [`DebugDropBomb`](https://docs.rs/drop_bomb/latest/drop_bomb/) to make sure that the consumer of the enter API don't forgot to call the `exit_stmt` / `exit_expr` methods which will defuse the bomb.